### DISCUSSION TITLE
Welcome to OrientationSensorFusion-ESP Discussions!

### DISCUSSION THREAD

**Discussion by BjarneBitscrambler** [BjarneBitscrambler/OrientationSensorFusion-ESP#16] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by DanielG86** [comment]
Hello Bjarne,
A bit busy with work related items lately, so little time to contribute, unfortunately.
I'm trying to implement a electronic compass on my boat, to replace the existing electronic compass, which is too slow for my radar overlay on my MFD. And implement some other sensors along the way with Signal K/ Openplotter.
When time permits, I can do some testing on an ESP32, or D1-mini. I'm not a real coder, but have a background in physics.

**Reply by BjarneBitscrambler** [reply]
Hey Daniel!
You're the 1st person to use this new GitHub feature - the Discussions.

Your reason for wanting an e-compass is the same as mine. We bought a Raymarine system last year, and while we like it a lot, I want to take advantage of features like the chart overlay you mention.

No problem about having only limited coding experience. As you've probably noticed from the Wiki and Code commits, things are looking pretty promising. I've finished porting the version 7 fusion code (which has built-in magnetic calibration, unlike version 4). Just yesterday I managed to fix a few problems that were preventing it from running on an ESP8266 (e.g. d1_mini), so now it seems fine on both that platform and the ESP32 processor. There are a few more tweaks I need to do to make it a proper Arduino library, which will probably take a week.  Then the last step is to integrate it into Signal K. Since I had already done that with the version 4 fusion code, it should be fairly straightforward. Anyway, I'll find it very helpful if you (and others) will compile and run the code on their own hardware and provide feedback on their experience to make the code more robust.

Despite the fusion code not being integrated yet with Signal K, it is still in a form that can be used. Instructions in the project Readme.md step through how to install the library. The sample program located in `examples/fusion_text_output.cc` will print out orientation data to a serial terminal.  All one needs is an ESP32 or ESP8266 platform connected to an orientation sensor like the Adafruit FXOS8700/FXAS21002 board.

Stay in touch.  Cheers,
Bjarne

**Comment by DanielG86** [comment]
I started with the project again, using a ESP32-Lolin board that I've used before. Couple of remarks:
The example code in "fusion_text_output.cc" gave me a compile error that a "}" was missing at the end. I've included it and then it compiles without problems.
And it works like a charm as well!
Didn't do any detailed testing of accuracy, but it looks like it is immediately calibrated (at least reasonably) from the output on the Serial Monitor. But of course this needs to be analyzed in more detail.

So, very good results. Thanks a lot for the good work. 
When you've integrated with Signal K, I can do some more testing, because I've set up the output graphs already from testing your earlier work.
Perhaps I have time this weekend to test with the D1-mini.

Great work so far!!

**Comment by DanielG86** [comment]
Also nice to have a temperature output. I wasn't aware of this sensor on the board. If possible, please include that output also in the Signal K integration. Can't have enough temperature sensors (although I'm unaware of the accuracy)

**Reply by BjarneBitscrambler** [reply]
Will do.  The spec sheet for the IC says that the temperature is uncalibrated, and on my board seems to be about 2 C lower than actual temperature. It does seem repeatable though, so calibrating it with an offset (which will probably vary between boards) should be straightforward,  I think Signal K/ESP has a provision for calibration offsets whereby they can be entered using the web interface to the sensor.

I've committed an example main.cc that corrects the mismatched { } you discovered.

**Reply by DanielG86** [reply]
Did a compile and upload on a D1 mini and it works like a charm. Fairly reasonable calibration directly from scratch, but I haven't tested it thoroughly, because the hardware is not properly setup yet. When you have the SensESP integration done, I can go in more depth, because I have graphs available for trends etc.
The good thing is that it looks there is no overshoot in the readings, and they are stable very quickly. So quite an improvement versus the original code.

**Comment by DanielG86** [comment]
Hi Bjarne,
I saw that you finished the SensESP integration. Is there a way that I could test the code? I'm not familiar with GitHub, so where can I find the files?

**Comment by BjarneBitscrambler** [comment]
Hi Daniel,
Merry Christmas, or any other seasonal celebration you have!  I hope all is
well with you.

Well, calling it finished is premature as there are still a few things I
want to address, but it *is* working and that's exciting. There are a
couple of ways you can bring in the code changes: manually, and using git.
I'll describe both as best I can, and you can decide.  If you can't get
something to work, then by all means let me know...

*Manual method.  *
Assuming you already have the working SensESP code on your local computer
(a recent version or the latest from the master branch), then you'll need
to copy over a half-dozen files from my repository: So browse over to
https://github.com/BjarneBitscrambler/SensESP/tree/OrientationV7 which
should show you my version of SensESP. Ensure that you are looking at the
"Code" tab, and that the "OrientationV7" branch is selected (circled in the
attached screenshot). You should then be able to search out the following
files:
- platformio.ini   Locate the "libdeps=" section and copy over the changes
from my file to the version of platformio.ini you are using. I've removed a
few of the unneeded AdaFruit libraries and added in the Orientation library
- examples/fx8700_heading_example.cpp  This is the example that you can
copy and rename to use as your main.cpp  The present version outputs
Attitude and CompassHeading Signal K messages
- src/sensors/orientation_sensor.cpp and orientation_sensor.h  Copy these
new files to your src/sensors directory
- /src/sensors/orientation_9dof_input.cpp and .h  Delete these two files
- /src/sensors/sensor_nxp_fxos8700_fxas21002.cpp and .h  Delete these two
files
- src/signalk/signalk_attitude.h
<https://github.com/BjarneBitscrambler/SensESP/blob/OrientationV7/src/signalk/signalk_attitude.h>
Copy this new file into your same directory
- src/signalk/signalk_output.h
<https://github.com/BjarneBitscrambler/SensESP/blob/OrientationV7/src/signalk/signalk_output.h>
Copy this changed file into your same directory

That should let you compile the new orientation code (assuming I haven't
forgotten any steps/files)

*Git method*
The second way is more automated and possibly easier. It requires that you
have command-line access to git on your computer. But I agree that git
takes some getting used-to...
- In a directory that you want to use for your project, type *git clone
https://github.com/BjarneBitscrambler/SensESP.git
<https://github.com/BjarneBitscrambler/SensESP.git>*  This should copy the
code to a new directory called SensESP.
- change to the new directory, and type *git checkout OrientationV7*  This
should change your working branch the directory to the OrientationV7 branch
(which is the one I am developing the orientation code on)
- copy the file examples/fx8700_heading_example.cpp  to src/main.cpp
This should now compile and run.

Note that whichever method you use, you may have to change platformio.ini
settings to select the platform you are compiling for (i.e. select d1_mini,
or esp32dev, or esp-wrover-kit

I may be repeating stuff you already know - or I may not be explaining in
enough detail. Either way, let me know how you progress! Cheers,
Bjarne

**Comment by DanielG86** [comment]
Hello Bjarne,

Merry Christmas and Happy Newyear to you as well!

Thanks a lot for this extensive reply. Very helpful indeed.
I actually couldn’t find your updated code, but with your reply this became clear. I dowloaded it all and copied the entire download in VS Code. With some minor adjustments for my board and usb-port, it all worked. I’ve got it now running on a D1mini and will test on a ESP32 later.
Output looks good. I think the sensor needs some initial calibration, which I did by turning the device in all directions a couple of times. Prior to this, there was a slow decline in the heading data, but now it looks stable. Will do some more testing, e.g. how quickly it responds, and whether there is overshoot in the readings.

Met vriendelijke groet,

Daniel Goedhuis
06-52 52 50 62


> Op 30 dec. 2020, om 19:48 heeft Bjarne Hansen <notifications@github.com> het volgende geschreven:
> 
> 
> Hi Daniel,
> Merry Christmas, or any other seasonal celebration you have! I hope all is
> well with you.
> 
> Well, calling it finished is premature as there are still a few things I
> want to address, but it *is* working and that's exciting. There are a
> couple of ways you can bring in the code changes: manually, and using git.
> I'll describe both as best I can, and you can decide. If you can't get
> something to work, then by all means let me know...
> 
> *Manual method. *
> Assuming you already have the working SensESP code on your local computer
> (a recent version or the latest from the master branch), then you'll need
> to copy over a half-dozen files from my repository: So browse over to
> https://github.com/BjarneBitscrambler/SensESP/tree/OrientationV7 which
> should show you my version of SensESP. Ensure that you are looking at the
> "Code" tab, and that the "OrientationV7" branch is selected (circled in the
> attached screenshot). You should then be able to search out the following
> files:
> - platformio.ini Locate the "libdeps=" section and copy over the changes
> from my file to the version of platformio.ini you are using. I've removed a
> few of the unneeded AdaFruit libraries and added in the Orientation library
> - examples/fx8700_heading_example.cpp This is the example that you can
> copy and rename to use as your main.cpp The present version outputs
> Attitude and CompassHeading Signal K messages
> - src/sensors/orientation_sensor.cpp and orientation_sensor.h Copy these
> new files to your src/sensors directory
> - /src/sensors/orientation_9dof_input.cpp and .h Delete these two files
> - /src/sensors/sensor_nxp_fxos8700_fxas21002.cpp and .h Delete these two
> files
> - src/signalk/signalk_attitude.h
> <https://github.com/BjarneBitscrambler/SensESP/blob/OrientationV7/src/signalk/signalk_attitude.h>
> Copy this new file into your same directory
> - src/signalk/signalk_output.h
> <https://github.com/BjarneBitscrambler/SensESP/blob/OrientationV7/src/signalk/signalk_output.h>
> Copy this changed file into your same directory
> 
> That should let you compile the new orientation code (assuming I haven't
> forgotten any steps/files)
> 
> *Git method*
> The second way is more automated and possibly easier. It requires that you
> have command-line access to git on your computer. But I agree that git
> takes some getting used-to...
> - In a directory that you want to use for your project, type *git clone
> https://github.com/BjarneBitscrambler/SensESP.git
> <https://github.com/BjarneBitscrambler/SensESP.git>* This should copy the
> code to a new directory called SensESP.
> - change to the new directory, and type *git checkout OrientationV7* This
> should change your working branch the directory to the OrientationV7 branch
> (which is the one I am developing the orientation code on)
> - copy the file examples/fx8700_heading_example.cpp to src/main.cpp
> This should now compile and run.
> 
> Note that whichever method you use, you may have to change platformio.ini
> settings to select the platform you are compiling for (i.e. select d1_mini,
> or esp32dev, or esp-wrover-kit
> 
> I may be repeating stuff you already know - or I may not be explaining in
> enough detail. Either way, let me know how you progress! Cheers,
> Bjarne
> 
> 
>

**Reply by BjarneBitscrambler** [reply]
Great that you got it working!
About the magnetic calibration: during boot-up the orientation library checks non-volatile memory for a valid calibration, and loads it if available. Then during normal operation, it periodically compares the existing magnetic field measurements to the existing calibration, and identifies whether a new calibration would be better than the existing calibration (if any).  I suspect your device hasn't previously had a calibration saved in EEPROM, so every time on boot-up it needs to gather enough readings to come up with a calibration (which is helped by rotating the device several times). Once it has generated a calibration, it should continue to work well until either the magnetic environment changes, or it is powered down.

Saving a magnetic calibration to EEPROM makes it so the sensor already is calibrated on boot-up. However the saving to EEPROM requires a command from the NXP Sensor Toolkit program (the one that runs under Windows). I'm going to implement a method in the SensESP code to cause the device to perform the save, without needing to use the NXP Windows toolkit.

Hope that makes sense.  Let me know about any other observations you have - this is helpful!  Cheers,
Bjarne

**Comment by Ince1956** [comment]
Hi would this be a good example to use as a basis for a sensesp gps?
